# MeChIP2: A Comprehensive ChIP-Seq Bioinformatics Pipeline

## What is MeCHIP2?
MeCHIP2 is an automated bioinformatics pipeline that downloads raw paired-end ChIP-seq data from the National Center for Biotechnology Information (NCBI) Sequence Read Archive (SRA) and uses it to generate the files required for data visualization as well as peak-calling. ChIP-seq data is downloaded directly from the SRA using the SRA Toolkit (sra-tools, v2.10.9). The raw sequence read file is then split into its respective paired-end read files by the SRA Toolkit. Paired-end reads are aligned to the GRCm39/mm39 mouse genome assembly using the Burrows-Wheeler Algorithm (BWA) Maximal Exact Match (MEM) software (bwa, v0.7.17). Samtools (samtools, v1.12) is used for file format conversion and deduplication of sequence data. Deeptools (deeptools, v3.5.1) is used to convert data to the bigwig file format, which can be used for visualization of ChIP-seq data in the UCSC Genome Browser or other visualization tools. 
[still need to add fastqc stuff and macs2 stuff]

## Using the Pipeline
In order to use the pipeline, clone the [GitHub repository](https://github.com/vhaghani26/MeChIP2.git) into the directory of your choice. The contents of the repository are broken down below for your reference. See the "Snakefile" section under "Repository Contents" for more specifics on file execution. It is  recommended to try the tutorial outlined at the end of this document to better understand specific usage of the program. Additionally, Snakemake should be installed in order to run the pipeline. See [here](https://snakemake.readthedocs.io/en/stable/getting_started/installation.html) for instructions on Snakemake installation.

### Repository Contents

#### samples.yaml
Open the `samples.yml` file and add each SRA ID you are interested in visualizing on its own line using the following format:
```
---
samples: 
   - SRAID1
   - SRAID2
```
This file will be used as a configuration file to pipe in the samples used in the analysis. The syntax in this file is important, so be aware that errors relating to `wildcards` are likely caused by the entries in this file.

#### Snakefile
The Snakefile contains the code required to execute the entire pipeline. Here, I will go through the basics of Snakemake usage for this program. For additional information, see the [Snakemake documentation](https://snakemake.readthedocs.io/en/stable/tutorial/tutorial.html). For specifics regarding command-line interface, [this](https://snakemake.readthedocs.io/en/stable/executing/cli.html) may be helpful.

Briefly, the whole program can be executed by running `snakemake -j 4 -p`, where the `-j` or `--jobs` option specifies the highest number of jobs (highest number of CPU) that can be run in parallel. The `-p` or `--prioritize` option tells the job scheduler to prioritize certain jobs given their dependencies. Without specifying anything else, the whole workflow will run.

To run a rule in particular, you can use `snakemake -j 4 -p rule_name`. The list of rules goes as follows:
* 


#### 00_conda_software
The repository contains a folder called `00_conda_software` that contains individual `yml` files with each specified software, including its version number, to be used for each rule in the pipeline. 
#### .gitignore
The `.gitignore` file contains the directories and consequently the data generated by each rule. Because the data is not pushed to git, it is strongly recommended to back up your data somewhere that it is protected. 
#### dag_mechip2.pdf
#### mechip2.slurm


[insert information about snakefile use here]

## Outputs
There will be four output directories, each containing a component of the pipeline.

[add more in-depth info here about what the file is]

### 01_raw_data
All sequence data for both the samples and reference genome, including reference genome alignment files, are stored in this directory.

### 02_sam_files
SAM files contains the SAM files generated for each sample by the reference genome alignment.

### 03_bam_files
All BAM files are stored in this folder, including intermediates of samtools flagging, sorting, and deduplication.

### 04_bigwig_files
Bigwig files are used for visualization of ChIP-seq data and are the final product of the pipeline.

## Tutorial
[write tutorial]
